src/utilities/modules/vaccines.js

```diff
@@ -0,0 +1,170 @@
+      }
+    });
+  });
+
+  // Push any remaining sensorlogs. If any are left, they form a breach
+  // with no delimiting non-breached sensorlog.
+  if (sensorLogStack.length > 0) breaches.push(sensorLogStack);
+
+  // Create Realm.result objects for each breach.
+  return breaches.map(breach => {
+    const sensorIds = breach.map(log => log.id);
+    const queryString = sensorIds.map((_, i) => `id = $${i}`).join(' OR ');
+    return database.objects('SensorLog').filtered(queryString, ...sensorIds);
+  });
+};
+
+/**
+ * Extracts all ItemBatches which are related to the passed sensorLogs.
+ * Groups all ItemBatch objects by Item and calculates the duration of
+ * time each ItemBatch was present during these logs.
+ *
+ * Option parameters itemBatch and item are used to filter the ItemBatches
+ * for these sensorLogs to only search for ItemBatch records which are
+ * related to the Item xor the ItemBatch. Item takes priority over
+ * ItemBatch if both are passed.
+ *
+ * Method iterates through each batch, grouping batches by Item. Will store
+ * values on the first encounter of a batch, otherwise will just increment
+ * the duration.
+ *
+ * @param  {Realm.results/array} sensorLogs
+ * @param  {ItemBatch}           itemBatch
+ * @param  {Item}                item
+ * @return {Array<Object>} array of objects, see example below.
+ * [
+ *  {
+ *     item,
+ *     batches: [ { id, code, expiryDate, enteredDate, duration, totalQuantity }]
+ *  }
+ *  ...
+ * ]
+ */
+export const sensorLogsExtractBatches = ({ sensorLogs = [], itemBatch, item } = {}) => {
+  const groupedBatches = {};
+  sensorLogs.forEach(({ itemBatches, logInterval, isInBreach }) => {
+    // Premature exit if log has no batches, or is not a breached sensorLog
+    if (itemBatches.length === 0 || !isInBreach) return;
+    // Ensure each batch has stock
+    let batchesToUse = itemBatches.filtered('totalQuantity > $0', 0);
+    // If an Item has been passed, only find batches for this item.
+    if (item) batchesToUse = batchesToUse.filtered('batch.id = $0', item.id);
+    // If an ItemBatch has been passed, only find batches for this item.
+    else if (itemBatch) batchesToUse = batchesToUse.filtered('batch.id = $0', itemBatch.id);
+    // For each batch, store it's details:
+    // { duration, id, code, enteredDate, totalQuantity, expiryDate }
+    // If already stored, increment the duration in the breach by this logs interval.
+    batchesToUse.forEach(
+      ({ item: batchItem, id: batchId, code, enteredDate, expiryDate, totalQuantity }) => {
+        // Premature return if this ItemBatch does not have an Item associated
+        if (!batchItem) return;
+        const { id: itemId } = batchItem;
+        // If this batches item hasn't been encountered yet, create
+        // a grouping object. groupedBatches = { itemId: { item, batches: {}  }}
+        if (!groupedBatches[itemId]) groupedBatches[itemId] = { item: batchItem, batches: {} };
```

Comment:
not sure why you would destrcuture item: to batchItem earlier

Reply:
parameter named item